python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -0,0 +1,530 @@
+            if request.prompt_token_ids is not None:
+                llm_prompt = self.vllm.inputs.data.TokensPrompt(
+                    prompt_token_ids=request.prompt_token_ids
+                )
+            else:
+                assert request.prompt
+                llm_prompt = self.vllm.inputs.data.TextPrompt(prompt=request.prompt)
+
+        # Send the request to the LLM engine.
+        stream = await self.engine.add_request(
+            request_id=str(request.request_id),
+            prompt=llm_prompt,
+            params=request.params,
+        )
+        # Consume the stream until the request is finished.
+        async for request_output in stream:
+            if request_output.finished:
+                # Bypass the original full prompt.
+                request_output.prompt = request.prompt
+                return request_output
+        raise RuntimeError("Should not reach here")
+
+    async def generate_async_v1(self, request: LLMRequest) -> Any:
+        """Process a single request.
+
+        Args:
+            request: The request.
+
+        Returns:
+            The output of the request.
+        """
+
+        # NOTE: vLLM v1 tighly couples tokenizer and detokenizer to the engine.
+        # We should investigate whether decoupling them could lead to better
+        # performance. Given that v1 tokenizer and detokenizer are already
+        # in a separate process, the benefit of decoupling them in the Processor
+        # may be limited.
+        assert request.prompt
+        multi_modal_data = {"image": request.images} if request.images else None
+        llm_prompt = self.vllm.inputs.data.TextPrompt(
+            prompt=request.prompt, multi_modal_data=multi_modal_data
+        )
+
+        # Send the request to the LLM engine.
+        stream = self.engine.generate(
+            request_id=str(request.request_id),
+            prompt=llm_prompt,
+            sampling_params=request.params,
+        )
+
+        # Consume the stream until the request is finished.
+        async for request_output in stream:
+            if request_output.finished:
+                # Bypass the original full prompt.
+                request_output.prompt = request.prompt
+                return request_output
+
+        raise RuntimeError("Should not reach here")
+
+    def shutdown(self):
+        """Shutdown the vLLM v1 engine. This kills child processes forked
+        by the vLLM engine. If not called, the child processes will be
+        orphaned and will not be killed when the parent process exits,
+        and they won't be able to be tracked by Ray anymore.
+        """
+        if hasattr(self.engine, "shutdown"):
+            logger.info("Shutting down vLLM engine")
+            self.engine.shutdown()
+
+
+class vLLMEngineStageUDF(StatefulStageUDF):
+    def __init__(
+        self,
+        data_column: str,
+        model: str,
+        engine_kwargs: Dict[str, Any],
+        task_type: str = "generate",
```

Comment:
make task_type enum

Reply:
Will do in a follow-up PR that introduce the vLLM processor, because this field needs to be propagated.